File: canva_pages_to_pdf.mjs
/**
 * Exporta un visor de Canva (u otro) a varios PDF, uno por “página”.
 * Avanza con la flecha derecha (típico en presentación).
 *
 * Uso:
 *   node canva_pages_to_pdf.mjs "<url>" <carpeta_salida> <num_paginas> [ms_espera]
 *
 * Ejemplo:
 *   node canva_pages_to_pdf.mjs "https://www.canva.com/design/.../view" ./canva_pdf 12 2000
 *
 * Notas:
 * - Se abre Chrome visible: inicia sesión en Canva si lo pide, entra al diseño y pulsa ENTER en la consola.
 * - Canva puede cambiar el DOM; si no avanza, ajusta la tecla o el tiempo de espera.
 * - Respeta los términos de uso de Canva; uso bajo tu responsabilidad.
 */

import fs from "fs/promises";
import path from "path";
import puppeteer from "puppeteer";

const sleep = (ms) => new Promise((r) => setTimeout(r, ms));

const [, , urlArg, outDirArg, pagesArg, delayArg] = process.argv;
if (!urlArg || !outDirArg || !pagesArg) {
  console.error(
    "Uso: node canva_pages_to_pdf.mjs \"<url>\" <carpeta_salida> <num_paginas> [ms_espera]"
  );
  process.exit(1);
}

const url = urlArg;
const outDir = path.resolve(outDirArg);
const totalPages = Math.max(1, parseInt(pagesArg, 10) || 1);
const delayMs = Math.max(500, parseInt(delayArg || "2000", 10) || 2000);

await fs.mkdir(outDir, { recursive: true });

const browser = await puppeteer.launch({
  headless: false,
  channel: "chrome",
  executablePath: "C:\\Program Files\\Google\\Chrome\\Application\\chrome.exe",
  defaultViewport: { width: 1920, height: 1080 },
  args: [
    "--disable-blink-features=AutomationControlled",
    "--start-maximized",
  ],
});

const page = await browser.newPage();
await page.goto(url, { waitUntil: "domcontentloaded", timeout: 120_000 });

process.stdout.write(
  "\n>>> Inicia sesión si hace falta, abre la vista correcta y deja visible la 1ª página.\n>>> Pulsa ENTER aquí para empezar a guardar PDFs...\n"
);
await new Promise((resolve) => process.stdin.once("data", resolve));

for (let i = 0; i < totalPages; i++) {
  const n = String(i + 1).padStart(3, "0");
  const filePath = path.join(outDir, `pagina_${n}.pdf`);

  await sleep(delayMs);
  await page.waitForNetworkIdle({ idleTime: 500, timeout: 10_000 }).catch(() => {});

  await page.pdf({
    path: filePath,
    printBackground: true,
    width: "1920px",
    height: "1080px",
    margin: { top: "0", right: "0", bottom: "0", left: "0" },
  });
  console.log(`Guardado: ${filePath}  (${i + 1}/${totalPages})`);

  if (i < totalPages - 1) {
    await page.keyboard.press("ArrowRight");
    await sleep(1500);
  }
}

await browser.close();
console.log("Listo.");

File: canva_pages_to_pdf.py
#!/usr/bin/env python3
"""
Exporta un visor de Canva (u otro) a varios PDF, uno por página.
Avanza con la flecha derecha.

Uso:
  pip install selenium webdriver-manager
  python canva_pages_to_pdf.py "<url>" <carpeta_salida> <num_paginas> [ms_espera]

Ejemplo:
  python canva_pages_to_pdf.py "https://www.canva.com/design/.../view" ./canva_pdf 12 2000

Notas:
- Chrome visible para poder iniciar sesión; luego Enter en consola.
- Canva puede cambiar el comportamiento; ajusta tiempo o teclas si falla.
"""

from __future__ import annotations

import argparse
import base64
import sys
import time
from pathlib import Path

from selenium import webdriver
from selenium.webdriver.chrome.options import Options
from selenium.webdriver.chrome.service import Service
from selenium.webdriver.common.by import By
from selenium.webdriver.common.keys import Keys
from webdriver_manager.chrome import ChromeDriverManager


def main() -> int:
    p = argparse.ArgumentParser(description="Canva / presentación → PDF por página (Selenium).")
    p.add_argument("url", help="URL del visor (p. ej. enlace view de Canva).")
    p.add_argument("out_dir", help="Carpeta de salida.")
    p.add_argument("pages", type=int, help="Número de páginas a capturar.")
    p.add_argument("delay_ms", type=int, nargs="?", default=2000, help="Espera antes de cada PDF (ms).")
    args = p.parse_args()

    if args.pages < 1:
        print("num_paginas debe ser >= 1", file=sys.stderr)
        return 1

    out = Path(args.out_dir).resolve()
    out.mkdir(parents=True, exist_ok=True)

    opts = Options()
    opts.add_argument("--window-size=1920,1080")
    opts.add_argument("--disable-blink-features=AutomationControlled")

    driver = webdriver.Chrome(
        service=Service(ChromeDriverManager().install()),
        options=opts,
    )
    try:
        driver.get(args.url)
        input(
            "\n>>> Inicia sesión si hace falta y deja la 1ª página visible.\n>>> Pulsa ENTER para empezar...\n"
        )

        body = driver.find_element(By.TAG_NAME, "body")

        for i in range(args.pages):
            time.sleep(max(0.5, args.delay_ms / 1000.0))
            pdf = driver.execute_cdp_cmd(
                "Page.printToPDF",
                {
                    "printBackground": True,
                    "paperWidth": 13.333,
                    "paperHeight": 7.5,
                    "marginTop": 0,
                    "marginBottom": 0,
                    "marginLeft": 0,
                    "marginRight": 0,
                },
            )
            raw = base64.b64decode(pdf["data"])
            path = out / f"pagina_{i + 1:03d}.pdf"
            path.write_bytes(raw)
            print(f"Guardado: {path}")
            if i < args.pages - 1:
                body.send_keys(Keys.RIGHT)

        print("Listo.")
        return 0
    finally:
        driver.quit()


if __name__ == "__main__":
    raise SystemExit(main())
